druid-shell/src/text_input.rs

```diff
@@ -0,0 +1,516 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::common_util::Counter;
+use crate::keyboard::{KbKey, KeyEvent};
+use crate::kurbo::{Point, Rect};
+use crate::piet::HitTestPoint;
+use crate::window::WinHandler;
+use std::borrow::Cow;
+use std::ops::Range;
+
+/// Uniquely identifies a text input field inside a window.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Hash)]
+pub struct TextInputToken(u64);
+
+impl TextInputToken {
+    /// A token that does not correspond to any text input.
+    pub const INVALID: TextInputToken = TextInputToken(0);
+
+    /// Create a new token; this should for the most part be called only by platform code.
+    pub fn next() -> TextInputToken {
+        static TEXT_FIELD_COUNTER: Counter = Counter::new();
+        TextInputToken(TEXT_FIELD_COUNTER.next())
+    }
+
+    /// Create a new token from a raw value.
+    pub const fn from_raw(id: u64) -> TextInputToken {
+        TextInputToken(id)
+    }
+
+    /// Get the raw value for a token.
+    pub const fn into_raw(self) -> u64 {
+        self.0
+    }
+}
+
+/// An event that indicates to the platform that some aspect of a text field may have changed.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+#[non_exhaustive]
+pub enum TextInputUpdate {
+    /// Indicates the value returned by `TextInputHandler::selection` may have changed.
+    SelectionChanged,
+
+    /// Indicates the values returned by one or more of these methods may have changed:
+    /// - `TextInputHandler::hit_test_point`
+    /// - `TextInputHandler::line_range`
+    /// - `TextInputHandler::bounding_box`
+    /// - `TextInputHandler::slice_bounding_box`
+    LayoutChanged,
+
+    /// Indicates any value returned from any of the TextInputHandler methods may have changed.
+    Reset,
+}
+
+/// Corresponds to a selection of some text. This also can represent a text input caret — the blinking vertical
+/// bar where text is to be inserted. A caret is a selection with zero length, where `anchor == extent`.
+/// Indices are always expressed in UTF-8 bytes, and must be between 0 and the document length, inclusive.
+///
+/// As an example, if the input caret is at the start of the document `hello world`, we would expect both `anchor`
+/// and `extent` to be `0`. If the user holds shift and presses the right arrow key five times, we would expect the
+/// word `hello` to be selected, the `anchor` to still be `0`, and the `extent` to now be `5`.
+#[derive(Clone, Default, PartialEq, Eq, Hash)]
+pub struct TextSelection {
+    /// This is the end of the selection that stays unchanged while holding shift and pressing the arrow keys.
+    pub anchor: usize,
+    /// This is the end of the selection that moves while holding shift and pressing the arrow keys.
+    pub extent: usize,
```

Comment:
I haven't seen these names before, are they your creation? I definitely can see wanting to avoid terms like 'start' and 'end', but I do find `extent` a bit unclear, since it feels like a measure of length or size, where it is really just another position. I'm not sure what I'd prefer though, but maybe `anchor` and `active`?

Reply:
[Chromium](https://source.chromium.org/chromium/chromium/src/+/master:content/browser/android/selection/selection_popup_controller.h;l=54-55;drc=659834c429b56ed3e9f6f2cc5a0636c393a72d5d) calls this base/extent, and I've seen anchor/focus in other places. Generally not partial to `focus`, since it's a different concept from focus in other UI contexts, but I don't mind `active`, can switch if nobody else has other opinions! I personally prefer `anchor` over `base` since an anchor suggests it won't move in some contexts, but also I've been told my naming suggestions have been bad in the past (originally on Fuchsia I proposed `anchor`/`head` which people hated) so probably best for me to defer to others' judgement :)

Reply:
one nice aspect of `extent` is it's a noun instead of an adjective, but that's a pretty minor nit

Reply:
I like anchor, I think that's clear. What's the natural inverse of `anchor`? `unmoored`? `adrift`? 🤡

extent.... just doesn't feel correct to me, because i equate extent with a measurement of some kind. Like: this would make sense to me if extent were an `isize`, relative to `anchor`. `anchor` and `edge`? `anchor` and `cursor`? 😱

Reply:
hahaa, the opposite of an anchor is maybe a `buoy`? a `sail`? a `raft`? i do kind of like `drift`, but it maybe falls into the same problem as extent where it sounds like a delta instead of an absolute value? also like buoy but maybe it's a little extravagant. also hard to spell.

i think `edge` is sometimes used to indicate either edge of the selection, so of all these options, maybe best to just stick with your original suggestion of active?